File: 112. Path Sum.py
from collections import deque

class Solution:
    def hasPathSum11(self, root: Optional[TreeNode], targetSum: int) -> bool:
        if not root: return False

        if root.val == targetSum and not root.left and not root.right:
           return True
        
        return self.hasPathSum(root.left, targetSum - root.val) or self.hasPathSum(root.right, targetSum - root.val)
    
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        if not root:
            return False
        
        stack = [(root, targetSum - root.val)]

        while stack:
            node, sum = stack.pop()

            if not node.left and not node.right and sum == 0:
                return True
            if node.left:
                stack.append((node.left, sum - node.left.val))
            if node.right:
                stack.append((node.right, sum - node.right.val))

        return False    

File: 112. Path Sum.go
func hasPathSum(root *TreeNode, targetSum int) bool {
    if root == nil{return false}

	if(root.Val == targetSum && root.Left == nil && root.Right == nil){
		return true
	}

	return hasPathSum(root.Left, targetSum - root.Val) || hasPathSum(root.Right, targetSum - root.Val)
}